breakdown/crawl4ai-llm-optimized-web-crawling-breakdown.md

```diff
@@ -0,0 +1,607 @@
+    media: Dict               # Images, videos, tables with metadata
+    links: Dict               # Internal/external links with scores
+
+    # Generated assets
+    screenshot: str           # Base64 encoded screenshot
+    pdf: bytes               # PDF representation
+    network_logs: List       # HTTP request/response logs
+```
+
+**Configuration Objects Hierarchy**
+
+```python
+# Browser-level configuration
+BrowserConfig:
+    headless: bool = True
+    user_data_dir: str = None
+    chrome_channel: str = "chrome"
+    browser_type: str = "chromium"
+
+# Per-crawl configuration
+CrawlerRunConfig:
+    cache_mode: CacheMode = CacheMode.ENABLED
+    extraction_strategy: ExtractionStrategy = NoExtractionStrategy()
+    session_id: str = None
+    word_count_threshold: int = 10
+    content_filter: ContentFilter = None
+```
+
+### Algorithms
+
+The content processing algorithms work together in a specific sequence to transform raw HTML into clean, AI-ready content:
+
+```mermaid
+flowchart TD
+    A[Raw HTML Content] --> B[WebScrapingStrategy Cleanup]
+    B --> C[PruningContentFilter]
+    C --> D{Content Filter Set?}
+
+    D -->|Yes| E[BM25ContentFilter]
+    D -->|No| F[DefaultMarkdownGenerator]
+    E --> F
+
+    F --> G[ExtractionStrategy]
+    G --> H{Strategy Type?}
+
+    H -->|LLM| I[LLMExtractionStrategy<br/>OpenAI/Anthropic/Ollama]
+    H -->|CSS| J[JsonCssExtractionStrategy<br/>CSS Selectors + Schema]
+    H -->|Regex| K[RegexExtractionStrategy<br/>Pattern Matching]
+
+    I --> L[Final CrawlResult]
+    J --> L
+    K --> L
+
+    L --> M[AdaptiveConfig Learning]
+    M --> N[SQLite Pattern Store]
+
+    subgraph "Content Quality Pipeline"
+        C
+        E
+        F
+    end
+
+    subgraph "Data Extraction Pipeline"
+        I
+        J
+        K
+    end
+
+    subgraph "Learning Pipeline"
+        M
+        N
+    end
+
+
+```
+
+**1. BM25 Content Filtering**
+
+The BM25 filter kicks in during content processing, right after the HTML gets cleaned up but before it becomes final markdown. When you give it a search query, Crawl4AI uses this to keep only the content that actually matches what you're looking for, which makes the output much more focused.
+
+```python
+class BM25ContentFilter:
+    def __init__(self, user_query: str, bm25_threshold: float = 1.0):
+        self.query_terms = user_query.lower().split()
+        self.threshold = bm25_threshold
+
+    def filter_content(self, content: str) -> str:
+        # Calculate BM25 scores for content chunks
+        # Filter chunks below threshold
+        # Return high-relevance content only
+```
+
+This runs when you set up the `content_filter` parameter in your crawler config. It happens after the basic HTML cleanup but before the final markdown gets generated. The filter breaks content into chunks and scores how well each chunk matches your query terms, then throws out anything that doesn't score high enough.
+
+**2. Strategy Pattern for Extraction**
```

Comment:
I think it’s a good idea to add a short introduction before showing just the code (or turn the code into bullet points)